homeassistant/components/sensor/__init__.py

```diff
@@ -508,17 +566,30 @@ def state(self) -> Any:
                 )
             return value
 
-        # If the sensor has neither a device class, a state class nor
-        # a unit_of measurement then there are no further checks or conversions
-        if not device_class and not state_class and not unit_of_measurement:
-            return value
+        precision = self.precision
 
-        if not self._invalid_numeric_value_reported and not isinstance(
-            value, (int, float, Decimal)
+        # If the sensor has neither a device class, a state class, a unit of measurement
+        # nor a precision then there are no further checks or conversions
+        if (
+            not device_class
+            and not state_class
+            and not unit_of_measurement
+            and precision is None
         ):
+            return value
+
+        # From here on a numerical value is expected
+        numerical_value: int | float | Decimal
+        if not isinstance(value, (int, float, Decimal)):
             try:
-                _ = float(value)  # type: ignore[arg-type]
+                if isinstance(value, str) and "." not in value:
+                    numerical_value = int(value)
+                else:
+                    numerical_value = float(value)  # type:ignore[arg-type]
             except (TypeError, ValueError):
+                # Raise if precision is not None, for other cases log a warning
+                if precision is not None:
+                    raise
```

Comment:
Should we raise a more explicit error?
This is what I was going to use in #85605
```python
            except (TypeError, ValueError) as err:
                raise ValueError(
                    f"Sensor {self.entity_id} has device class {device_class}, "
                    f"state class {state_class} and unit {unit_of_measurement} "
                    "thus indicating it has a numeric value; however, it has the "
                    f"non-numeric value: {value} ({type(value)})"
                ) from err
```

Reply:
I don't think that's needed now since it's a new property. Once this is merged, you can rebase #85605 to make the error message more elaborate also if `native_precision` is not `None` and the sensor is not numeric.

Reply:
Not even a simple version?
```python
                raise ValueError(
                    f"Couldn't convert non-numeric value: {value} ({type(value)})"
                ) from err
```

Reply:
I actually think it is nice to provide a little more context to the error message.
It makes it much more understandable what is wrong (when it goes south).